cpu/atmega2560/periph/gpio.c

```diff
@@ -0,0 +1,244 @@
+        ret_code = bit_is_clear(_SFR_MEM8(_ddr_addr(pin)), _pin_num(pin));
+    }
+
+    if (ret_code == 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
+                  gpio_cb_t cb, void *arg)
+{
+    uint8_t pin_num = _pin_num(pin);
+    int ret_code = gpio_init(pin, GPIO_DIR_IN, pullup);
+
+    if (ret_code < 0) {
+        return ret_code;
+    }
+
+    gpio_set(pin);
+
+    /* clear global interrupt flag */
+    cli();
+
+    EIMSK |= (1 << pin_num);
+
+    /* configure the flank */
+    switch (flank) {
+        case GPIO_RISING:
+            if (pin_num < 4) {
+                EICRA |= (3 << pin_num * 2);
+            }
+            else {
+                EICRB |= (3 << (pin_num * 2) % 4);
+            }
+            break;
+        case GPIO_FALLING:
+            if (pin_num < 4) {
+                EICRA |= (2 << pin_num * 2);
+            }
+            else {
+                EICRB |= (2 << (pin_num * 2) % 4);
+            }
+            break;
+        case GPIO_BOTH:
+            if (pin_num < 4) {
+                EICRA |= (1 << pin_num * 2);
+            }
+            else {
+                EICRB |= (1 << (pin_num * 2) % 4);
+            }
+            break;
+        default:
+            return -1;
+    };
+
+    /* set callback */
+    config[pin_num].cb = cb;
+    config[pin_num].arg = arg;
+
+    /* set global interrupt flag */
+    sei();
+
+    return 0;
+}
+
+void gpio_irq_enable(gpio_t pin)
+{
+    EIMSK |= (1 << _pin_num(pin));
+}
+
+void gpio_irq_disable(gpio_t pin)
+{
+    EIMSK &= ~(1 << _pin_num(pin));
+}
+
+int gpio_read(gpio_t pin)
+{
+    if (bit_is_set(_SFR_MEM8(_port_addr(pin)), _pin_num(pin)) > 0) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void gpio_set(gpio_t pin)
+{
+    _SFR_MEM8(_port_addr(pin)) |= (1 << _pin_num(pin));
+}
+
+void gpio_clear(gpio_t pin)
+{
+    _SFR_MEM8(_port_addr(pin)) &= ~(1 << _pin_num(pin));
+}
+
+void gpio_toggle(gpio_t pin)
+{
+    if (gpio_read(pin)) {
+        gpio_clear(pin);
+    }
+    else {
+        gpio_set(pin);
+    }
+}
+
+void gpio_write(gpio_t pin, int value)
+{
+    if (value) {
+        gpio_set(pin);
+    }
+    else {
+        gpio_clear(pin);
+    }
+}
+
+static inline void irq_handler(uint8_t pin_num)
+{
+    config[pin_num].cb(config[pin_num].arg);
+}
+
+ISR(INT0_vect, ISR_BLOCK)
+{
+    irq_handler(_pin_num(GPIO(PORT_D, 0))); /**< predefined pin */
+}
+
+ISR(INT1_vect, ISR_BLOCK)
+{
+    irq_handler(_pin_num(GPIO(PORT_D, 1))); /**< predefined pin */
+}
+
+ISR(INT2_vect, ISR_BLOCK)
+{
+    irq_handler(_pin_num(GPIO(PORT_D, 2))); /**< predefined pin */
+}
+
+ISR(INT3_vect, ISR_BLOCK)
+{
+    irq_handler(_pin_num(GPIO(PORT_D, 3))); /**< predefined pin */
+}
+
+ISR(INT4_vect, ISR_BLOCK)
+{
+    irq_handler(_pin_num(GPIO(PORT_E, 4))); /**< predefined pin */
+}
+
+ISR(INT5_vect, ISR_BLOCK)
+{
+    irq_handler(_pin_num(GPIO(PORT_E, 5))); /**< predefined pin */
+}
```

Comment:
Currently I'm thinking if we could make this a bit prettier. But as we don't define those vectors...